### PR TITLE
add weekOfMonth

### PR DESCRIPTION
**Pull request by suisha**
Implements https://github.com/moment/moment/issues/2934

Some questions that occurred to me while working on this:
- Does it make sense to implement this function as a standalone method vs integrating it into `moment.week`?
- Do we also need an `ISO` version of this method, similar to `isoWeek`?
- Are my test cases sufficient?

### CODE REVIEW

File: src/test/moment/weeks.js [hunk 1]
```diff
@@ -118,6 +118,14 @@ test('weeks setter', function (assert) {
     assert.equal(moment([2012, 0, 15]).week(30).week(), 30, 'Setting Jan 15 2012 to week 30 should work');
 });
 
+test('weekOfMonth', function (assert) {
+    assert.equal(moment([2012, 0,  1]).weekOfMonth(), 1, 'Setting Jan 1 2012 to week 30 should work');
+    assert.equal(moment([2012, 0,  7]).weekOfMonth(), 1, 'Setting Jan 7 2012 to week 30 should work');
+    assert.equal(moment([2012, 0,  8]).weekOfMonth(), 2, 'Setting Jan 8 2012 to week 30 should work');
+    assert.equal(moment([2012, 0, 14]).weekOfMonth(), 2, 'Setting Jan 14 2012 to week 30 should work');
+    assert.equal(moment([2012, 0, 15]).weekOfMonth(), 3, 'Setting Jan 15 2012 to week 30 should work');
```
Comment (icambron) [inline src/test/moment/weeks.js]: These descriptions aren't quite right; they should be phrased more like "Jan 15 2012 is the third week of January".

**Comment by icambron** [comment]
Except that one detail in the tests, this looks good to me, @ichernev

File: src/test/moment/weeks.js (hunk 1 above)
Comment (suisha) [inline src/test/moment/weeks.js]: Fixed :hammer:

**Comment by mj1856** [comment]
LGTM. Staged for next release.  Thanks!

**Comment by ichernev** [comment]
Well, the logic that says which is the first week in a given year is a) locale specific b) not so simple. I'd rather use that when deciding what is the first week of the month, rather than first day of month defines first week of month.

@icambron @mj1856 what do you think?

**Comment by icambron** [comment]
I guess I don't understand how week-years can be used in calculating week-of-month

**Comment by ichernev** [comment]
For week-of-year you take into account which day (of the year) defines the first week. ISO says its Thursday, so if new year starts in Friday (Sat or Sun), its still last week-year, if it starts Monday, Tuesday, Wednesday, Thursday, it will be the first week. Different start of week (Mon vs Sat vs Sun) add a bit of complexity on top.

Locales that don't use ISO use first day-of-year to define the first week, which is easier for computation. But its just per locale setting, so we need the (harder) computation anyway.

Same (I think) should apply for months. 

I spend some time during the last release to rework the week-related computations, so now at least they make sense to me (and hopefully to others). Take a look at the math (and docs) in this file:

https://github.com/moment/moment/blob/develop/src/lib/units/week-calendar-utils.js

**Comment by icambron** [comment]
Oh I see. Sure, that makes sense.

**Comment by ichernev** [comment]
@suisha can you use the logic from here: https://github.com/moment/moment/blob/develop/src/lib/units/week-calendar-utils.js#L40 to implement that monthWeek. Basically take into account locale dow/doy to compute the weekOffset (repositions you to the begining of the week), so computations make sense.

File: src/lib/units/week.js [hunk 2]
```diff
@@ -55,6 +55,10 @@ export function getSetWeek (input) {
     return input == null ? week : this.add((input - week) * 7, 'd');
 }
 
+export function getWeekOfMonth (input) {
```
Comment (OlsonDev) [inline src/lib/units/week.js]: Unused variable `input`

**Comment by maggiepint** [comment]
I think there's a lot of interest in this, but it isn't getting done. What do we need to do to move this forward? I think perhaps instructions are unclear.

**Comment by suisha** [comment]
Hey, sorry, I totally forgot about this PR.

I'll take a stab at this by the end of this week. If that's too slow, feel free to assign this to someone else. Didn't mean to keep people waiting.

**Comment by suisha** [comment]
As promised, I took another try at this PR, and before I implement I want to make sure I understand the issues here. This is my understanding of @ichernev 's comments:

1. To determine week of year, we need to first determine locale.
 - If UTC, it's the first week with a Thursday
 - If it's any other locale, it's whichever week contains the first day of the year.

2. Under the current implementation of `weekOfMonth` in this PR, the first week of the month will always be whichever week contains the first day of the month. This is incorrect, because the week of month should also be locale specific, like week of year.

Therefore, the expected behavior for this function should be:
  - If UTC, the first week of the month is the first week with a Thursday
  - If any other locale, use the week with the first day of the month

...which is basically the same as week of year.

If my summary above is correct, then I have the following concern:
> [The ISO standard does not define any association of weeks to months. A date is either expressed with a month and day-of-the-month, or with a week and day-of-the-week, never a mix.](https://en.wikipedia.org/wiki/ISO_week_date#Weeks_per_month)

I understand this to mean that there is no standard definition of the term `weekOfMonth`. Therefore, the new functionality we are proposing is arbitrary and not according to any standard. Is it really necessary to make this feature locale specific?

**Comment by ichernev** [comment]
@suisha thank you for the detailed question, we rarely get these here :)

You're correct in that there is UTC and first day rules, and that there is no rule for `weekOfMonth`. 

However, we try to follow principle of least surprise. So if we have rules of determining week of year, I don't see why not apply same principle for week of month.

Also, sometimes people use locales with non-standard dow/doy combinations (and they are moment users): https://github.com/moment/moment/pull/2336 So the code should not assume that whatever is defined in the current locales is all and complete.

Also the logic that picks the week with the generic (following dow/doy rules) is rather simple: https://github.com/moment/moment/blob/develop/src/lib/units/week-calendar-utils.js#L40-41

**Comment by suisha** [comment]
@ichernev Can you explain what `firstWeekOffset` is calculating? 

My locale has `dow = 0` and `doy = 6`. Since every week contains a Saturday (it's the last day of the week), I expected the offset to be `0` (since the week with Jan 1 will always have a Saturday), but instead it's `-5`.

**Comment by suisha** [comment]
Hey, I apologize for taking so long on this. I unfortunately don't have the bandwidth to work on this anymore, so I'll go ahead and close. If someone wants to take this over again later they can re-open.

**Comment by Aides359** [comment]
After waiting for this feature I now take another go at this problem.
This might be an [interesting read](http://www.linguistlist.org/issues/12/12-1858.html) (it is some kind of survey on that topic to different people across the world).

**Pull request closed without merging**